### PR TITLE
Add call function

### PR DESCRIPTION
**Pull request by axelchalon**
Open for review; we might want to wait until we refactor the Parity updater before we merge this branch (or maybe not?), as some new issues might come up with the implementation.

### CODE REVIEW

**Comment by debris** [comment]
let's wait for refactor

**Comment by axelchalon** [comment]
I had to add `#[allow(dead_code)]` in the use_contract macro because otherwise the compiler would complain (in the parity updater) about unused methods in the derived contract. Is it normal that it should complain about unused methods of a module?

```
warning: method is never used: `output`                                                                                                                                  
  --> updater/src/operations.rs:10:1                                                                                                                                     
   |                                                                                                                                                                     
10 | use_contract!(operations_contract,"Operations","./res/operations.abi");                                                                                             
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                             
   |                                                                                                                                                                     
   = note: #[warn(dead_code)] on by default                                                                                                                              
   = note: this error originates in a macro outside of the current crate                                                                                                 
                                                                                                                                                                         
warning: method is never used: `call`                                                                                                                                    
  --> updater/src/operations.rs:10:1                                                                                                                                     
   |                                                                                                                                                                     
10 | use_contract!(operations_contract,"Operations","./res/operations.abi");                                                                                             
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                             
   |                                                                                                                                                                     
   = note: #[warn(dead_code)] on by default                                                                                                                              
   = note: this error originates in a macro outside of the current crate                                                                                                 
                                                                                                                                                                         
warning: method is never used: `checksum`                                                                                                                                
  --> updater/src/operations.rs:10:1                                                                                                                                     
   |                                                                                                                                                                     
10 | use_contract!(operations_contract,"Operations","./res/operations.abi");                                                                                             
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                             
   |                                                                                                                                                                     
   = note: #[warn(dead_code)] on by default                                                                                                                              
   = note: this error originates in a macro outside of the current crate                                                                                                 
                                                                                                                                                                         
warning: method is never used: `set_client_required`                                                                                                                     
  --> updater/src/operations.rs:10:1                                                                                                                                     
   |                                                                                                                                                                     
10 | use_contract!(operations_contract,"Operations","./res/operations.abi");                                                                                             
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                             
   |                                                                                                                                                                     
   = note: #[warn(dead_code)] on by default                                                                                                                              
   = note: this error originates in a macro outside of the current crate

etc.
```

**Comment by axelchalon** [comment]
PR is ready for review/merge

File: derive/src/lib.rs [hunk 1]
```diff
@@ -613,17 +613,13 @@ fn declare_functions(function: &Function) -> quote::Tokens {
...
-				// convert to ethabi::Result<ethabi::Bytes>
-				// match call(encoded_input) {
-				// 	Result::Ok(x) => ethabi::Result::Ok(x),
-				// 	Result::Err(x) => ethabi::Result::Err(x)
-				// }.and_then(|encoded_output| self.output(&encoded_output))
-
-				call(encoded_input).map_err(|x|	ethabi::Error::with_chain(ethabi::Error::from(x), ethabi::ErrorKind::CallError)).and_then(|encoded_output| self.output(&encoded_output))
+				do_call(encoded_input)
+					.map_err(|x| ethabi::Error::with_chain(x, ethabi::ErrorKind::CallError))
+					.and_then(|encoded_output| self.output(&encoded_output))
```
Comment (debris) [inline derive/src/lib.rs]: is it possible to simplify it and call only

```rust
do_call(encoded_input)
  .chain_err(|| ethabi::ErrorKind::CallError)
  .and_then(|encoded_output| self.output(&encoded_output))
```

File: derive/src/lib.rs [hunk 2]
```diff
@@ -612,6 +612,15 @@ fn declare_functions(function: &Function) -> quote::Tokens {
...
 				#o_impl
 			}
+
+			pub fn call<#(#template_params),*>(&self, #(#params ,)* do_call: &Fn(ethabi::Bytes) -> Result<ethabi::Bytes, String>) -> ethabi::Result<#output_kinds>
+			{
+				let encoded_input = self.input(#(#names),*);
+
+				do_call(encoded_input)
+					.map_err(|x| ethabi::Error::with_chain(ethabi::Error::from(x), ethabi::ErrorKind::CallError))
+					.and_then(|encoded_output| self.output(&encoded_output))
```
Comment (debris) [inline derive/src/lib.rs]: is it possible to simplify it and call only

```rust
do_call(encoded_input)
  .chain_err(|| ethabi::ErrorKind::CallError)
  .and_then(|encoded_output| self.output(&encoded_output))
```

**Comment by debris** [comment]
lgtm, @tomusdrw what do you think about it?

**Comment by debris** [comment]
@tomusdrw imo, this could be a separate feature. Currently the bridge does not depend on futures and I think it's useful in many places - especially our legacy code

**Comment by axelchalon** [comment]
@debris could you bump the version and update the package on crates.io so that we can use a regular crates.io dependency in the Parity updater PR?

**Comment by debris** [comment]
sure!
btw. everyone from `paritytech/core-devs` team can do that ;)

**Pull request merged**